language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/util/ForeachCompletionUtil.java

```diff
@@ -183,8 +185,15 @@ public static String getTypeOfIteratorVariable(BallerinaCompletionContext ctx, T
         TypeSymbol rawType = CommonUtil.getRawType(symbol);
         switch (rawType.typeKind()) {
             case STRING:
+                type = CommonUtil.getModifiedTypeName(ctx, rawType) + ":Char";
+                break;
             case XML:
-                type = CommonUtil.getModifiedTypeName(ctx, rawType);
+                Optional<TypeSymbol> xmlTypeParam = ((XMLTypeSymbol) symbol).typeParameter();
+                if (xmlTypeParam.isEmpty() || xmlTypeParam.get().typeKind() == TypeDescKind.UNION) {
```

Comment:
Why do we need to special case the union here?
I believe it is for the following,

```
function test(xml<xml:Element | xml:Comment> xx) {
    foreach xml:Element | xml:Comment item in xx {

    }
}
```

Even in that case, it is okay to check the empty type-parameter and get the modified type parameter in other cases.

Reply:
If we have the following 
  xml x1 = xml `<book>The Lost World</book>`;

we get the type-parameter as follows.

xml:Element|xml:Comment|xml:ProcessingInstruction|xml:Text

adding this union would be not suitable IMHO.

Reply:
Basically we don't need to do this explicitly, and when we consider the language specification as well as the semantic API, the type parameter is optional. In the sense we should show the type parameters only when the user has added it in the source. If the user has added all four elements in the source, we can show it (which is a rare case).